### PR TITLE
Throw UnsupportedOperationException from unimplemented Neo4jStatement methods

### PR DESCRIPTION
**Pull request by britter**
Fix for #17. Neo4jStatement should not fail silently when methods are called which are currently not supported.

I've added code that throws UOE from every method where it was obvious to me. There some methods I'm not sure about:
- execute(String, int)
- execute(String, int[])
- execute(String, String[])
- executeUpdate(String)
- executeUpdate(String, int)
- executeUpdate(String, int[])
- executeUpdate(String, String[])
- isWrapperFor(Class<?>)
- getMoreResults()
- getMoreResults(int)
- unwrap(Class<T>)

These methods contain logic, but they don't look correct to me. For example all the execute methods delegate to execute(String) and ignore any additional parameters.

### CODE REVIEW

**Comment by britter** [comment]
@jexp  any comments here?

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 1]
```diff
@@ -167,7 +173,7 @@ public ResultSet getResultSet() throws SQLException
     @Override
     public int getUpdateCount() throws SQLException
     {
-        return -1;
+        throw unsupported( "getUpdateCount" );
```
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:179]: we could actually return it if the last operation was an update

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 2]
```diff
@@ -71,44 +71,49 @@ public void close() throws SQLException
...
     @Override
     public void setQueryTimeout( int i ) throws SQLException
     {
+        throw unsupported( "setQueryTimeout" );
     }
 
     @Override
     public void cancel() throws SQLException
     {
+        throw unsupported( "cancel" );
```
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:118]: does this mean rollback on a statement level? would that also roll back the tx?

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 3]
```diff
@@ -180,29 +186,31 @@ public boolean getMoreResults() throws SQLException
...
     public void setFetchDirection( int i ) throws SQLException
     {
+        throw unsupported( "setFetchDirection" );
     }
 
     @Override
     public int getFetchDirection() throws SQLException
     {
-        return 0;
+        throw unsupported( "getFetchDirection" );
```
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: should it always return FD.FORWARD?

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 4]
```diff
@@ -180,29 +186,31 @@ public boolean getMoreResults() throws SQLException
...
     {
-        return 0;
+        throw unsupported( "getFetchSize" );
     }
 
     @Override
     public int getResultSetConcurrency() throws SQLException
     {
-        return 0;
+        throw unsupported( "getResultSetConcurrency" );
```
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: there should be a constant to be returned for not-concurrent.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 5]
```diff
@@ -214,17 +222,19 @@ public int getResultSetType() throws SQLException
     @Override
     public void addBatch( String s ) throws SQLException
     {
+        throw unsupported( "addBatch" );
```
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:228]: I have to implement this one :)

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 6]
```diff
@@ -296,6 +306,7 @@ public boolean isClosed() throws SQLException
     @Override
     public void setPoolable( boolean b ) throws SQLException
     {
+        throw unsupported( "setPoolable" );
```
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:312]: interesting that this is not on the connection level.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 7]
```diff
@@ -318,10 +329,16 @@ public boolean isWrapperFor( Class<?> aClass ) throws SQLException
 
     public void closeOnCompletion() throws SQLException
     {
+        throw unsupported( "closeOnCompletion" );
```
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:335]: I assume this sets a boolean flag? Which we could honor and finish the tx (either rollback or commit) after statement completion? which is probably when the resultset is exhausted?

File: src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java [hunk 8]
```diff
@@ -67,14 +68,14 @@ public void testExecuteStatement() throws Exception
         assertEquals( nodeId, ((Number) rs.getObject( "id" )).intValue() );
         assertEquals( nodeId, rs.getLong( "id" ) );
         assertEquals( nodeId, ((Number) rs.getObject( 1 )).intValue() );
-        assertEquals( nodeId, rs.getLong( 1 ) );
+        assertEquals(nodeId, rs.getLong(1));
```
Comment (jexp) [inline src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java]: probably don't change the code-style here.

File: src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java [hunk 9]
```diff
@@ -160,4 +161,125 @@ public void testInvalidColumnName() throws Exception
         rs.next();
         rs.getObject( "foo" );
     }
+
```
Comment (jexp) [inline src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java:163]: I think it makes sense to move the USO tests into a separate test?

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 2 above)
Comment (britter) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:118]: I don't know for sure, but [this discussion](http://stackoverflow.com/questions/295920/how-can-i-abort-a-running-jdbc-transaction) indicates that only the statement is canceled. `Connection.rollback()` has to be called explicitly after a statement has been canceled.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 1 above)
Comment (britter) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:179]: Okay, but how do we determine the update count? In a relational DB it simply the number of rows which have been touched. What do we count in Neo4j? Nodes updated? Nodes and edges updated? Properties updated?

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 3 above)
Comment (britter) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: Not sure if a fetch direction makes sense in context of a graph database. Maybe `ResultSet.FETCH_UNKNOWN` is a better fit?

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 4 above)
Comment (britter) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: `ResultSet` defines two constants: `CONCUR_READ_ONLY` and `CONCUR_UPDATABLE`. I guess we should return the first then?

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 5 above)
Comment (britter) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:228]: Okay, I'll leave it the way it is now and you just change the implementation when you have the time.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 7 above)
Comment (britter) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:335]: The JavaDoc says: 

> Specifies that this Statement will be closed when all its dependent result sets are closed. If execution of the Statement does not produce any result sets, this method has no effect.
> 
> **Note:** Multiple calls to closeOnCompletion do not toggle the effect on this Statement. However, a call to closeOnCompletion does effect both the subsequent execution of statements, and statements that currently have open, dependent, result sets.

So we would have to call close() on the Statement when close has been called on all result sets. I think this should be implemented in a different PR.

File: src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java (hunk 8 above)
Comment (britter) [inline src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java]: Sorry :o)

File: src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java (hunk 9 above)
Comment (britter) [inline src/test/java/org/neo4j/jdbc/Neo4jStatementTest.java:163]: Done!

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java [hunk 10]
```diff
@@ -318,10 +329,16 @@ public boolean isWrapperFor( Class<?> aClass ) throws SQLException
...
     }
 
     public boolean isCloseOnCompletion() throws SQLException
     {
         return false;
     }
+
+    private static UnsupportedOperationException unsupported( String methodName )
```
Comment (britter) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: I just got aware of `java.sql.SQLFeatureNotSupportedException`, maybe it is better to throw this exception. Note that it extends SQLException. So it can only be thrown from methods which declare SQLException in their throw clause.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 1 above)
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:179]: Great question. Relational databases count rows, so in our case that would man nodes + relationships updated. Perhaps we should at least add a //TODO here.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 3 above)
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: Isn't it about the order of the resultset? I.e. that you can go forward or backward? But UNKNOWN is fine for me too. Not sure if there are any tools relying on it.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 4 above)
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: Yep.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 5 above)
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:228]: Yep, thanks!

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 7 above)
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java:335]: Yes, so I add a GH issue for it. Thanks.

File: src/main/java/org/neo4j/jdbc/Neo4jStatement.java (hunk 10 above)
Comment (jexp) [inline src/main/java/org/neo4j/jdbc/Neo4jStatement.java]: Ok, makes sense. The Javadoc states:

This indicates that the JDBC driver does not support an optional JDBC feature. Optional JDBC features can fall into the fallowing categories:
- no support for an optional feature
- no support for an optional overloaded method
- no support for an optional mode for a method. 

The mode for a method is determined based on constants passed as parameter values to a method

**Comment by britter** [comment]
Do you want me to fix anything else before this can be integrated? What about the various execute overloads?

**Comment by jexp** [comment]
Sorry, total disconnect over the holidays :) Thanks a lot again for your work @britter
